### PR TITLE
Add table for VPC Network ACL. Closes #41

### LINKED ISSUE #41: Add table for VPC Network ACL

### PR DESCRIPTION
**Pull request by rajeshbal65**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT 300

SETUP: tests/alicloud_vpc_network_acl []

PRETEST: tests/alicloud_vpc_network_acl

TEST: tests/alicloud_vpc_network_acl
Running terraform
data.alicloud_caller_identity.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
alicloud_vpc.named_test_resource: Creating...
alicloud_vpc.named_test_resource: Creation complete after 6s [id=vpc-a2dou9votier9bw0gxhvp]
alicloud_network_acl.named_test_resource: Creating...
alicloud_network_acl.named_test_resource: Creation complete after 1s [id=nacl-a2dfcksumb2cf06zzssrk]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

account_id = ****************
network_acl_id = nacl-a2dfcksumb2cf06zzssrk
region_id = ap-south-1
resource_aka = acs:vpc:ap-south-1:****************:network-acl/nacl-a2dfcksumb2cf06zzssrk
resource_name = turbottest95569
vpc_id = vpc-a2dou9votier9bw0gxhvp

Running SQL query: test-get-query.sql
[
  {
    "description": "Test network acl table",
    "name": "turbottest95569",
    "network_acl_id": "nacl-a2dfcksumb2cf06zzssrk",
    "region": "ap-south-1",
    "vpc_id": "vpc-a2dou9votier9bw0gxhvp"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "description": "Test network acl table",
    "name": "turbottest95569",
    "network_acl_id": "nacl-a2dfcksumb2cf06zzssrk",
    "region": "ap-south-1",
    "vpc_id": "vpc-a2dou9votier9bw0gxhvp"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "account_id": "****************",
    "akas": [
      "acs:vpc:ap-south-1:****************:network-acl/nacl-a2dfcksumb2cf06zzssrk"
    ],
    "name": "turbottest95569",
    "title": "turbottest95569"
  }
]
✔ PASSED

POSTTEST: tests/alicloud_vpc_network_acl

TEARDOWN: tests/alicloud_vpc_network_acl

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```

### Basic info

```sql
select
  name,
  network_acl_id,
  status,
  vpc_id,
  description,
  region_id
from
  alicloud_vpc_network_acl;
+------------+----------------------------+-----------+---------------------------+---------------+------------+
| name       | network_acl_id             | status    | vpc_id                    | description   | region     |
+------------+----------------------------+-----------+---------------------------+---------------+------------+
| test-nacl1 | nacl-a2d4chqwh8znl3m10z24g | Available | vpc-a2dtp41aajqffw2vmw2vo | Test resource | ap-south-1 |
+------------+----------------------------+-----------+---------------------------+---------------+------------+
```

### List the VSwitch associated with network ACLs

```sql
select
  network_acl_id,
  vpc_id,
  association ->> 'ResourceId' as vswitch_id,
  association ->> 'Status' as association_status
from
  alicloud_vpc_network_acl,
  jsonb_array_elements(resources) as association
where
  association ->> 'ResourceType' = 'VSwitch';

+----------------------------+---------------------------+---------------------------+--------------------+
| network_acl_id             | vpc_id                    | vswitch_id                | association_status |
+----------------------------+---------------------------+---------------------------+--------------------+
| nacl-a2dcnd0dgkyzyydwusa15 | vpc-a2db7sc9m92c8gu3a1qw8 | vsw-a2dm9m53xiw9xl2176wdn | BINDED             |
+----------------------------+---------------------------+---------------------------+--------------------+
```

### List the inbound rule info of network ACLs

```sql
select
  name,
  network_acl_id,
  vpc_id,
  i ->> 'NetworkAclEntryId' as network_acl_entry_id,
  i ->> 'NetworkAclEntryName' as network_acl_entry_name,
  i ->> 'Description' as description,
  i ->> 'EntryType' as entry_type,
  i ->> 'Policy' as policy,
  i ->> 'Port' as port,
  i ->> 'Protocol' as protocol,
  i ->> 'SourceCidrIp' as source_cidr_ip
from
  alicloud_vpc_network_acl,
  jsonb_array_elements(ingress_acl_entries) as I;

+------+----------------------------+---------------------------+---------------------------+------------------------+-------------+------------+--------+-------+----------+----------------+
| name | network_acl_id             | vpc_id                    | network_acl_entry_id      | network_acl_entry_name | description | entry_type | policy | port  | protocol | source_cidr_ip |
+------+----------------------------+---------------------------+---------------------------+------------------------+-------------+------------+--------+-------+----------+----------------+
| test | nacl-a2dcnd0dgkyzyydwusa15 | vpc-a2db7sc9m92c8gu3a1qw8 | nae-a2dayapck66ngi2gxiimi |                        |             |            | accept | -1/-1 | all      | 0.0.0.0/0      |
+------+----------------------------+---------------------------+---------------------------+------------------------+-------------+------------+--------+-------+----------+----------------+

```
### List the outbound rule info of network ACLs

```sql
select
  name,
  network_acl_id,
  vpc_id,
  i ->> 'NetworkAclEntryId' as network_acl_entry_id,
  i ->> 'NetworkAclEntryName' as network_acl_entry_name,
  i ->> 'Description' as description,
  i ->> 'EntryType' as entry_type,
  i ->> 'Policy' as policy,
  i ->> 'Port' as port,
  i ->> 'Protocol' as protocol,
  i ->> 'DestinationCidrIp' as destination_cidr_ip
from
  alicloud_vpc_network_acl,
  jsonb_array_elements(egress_acl_entries) as I;

+------+----------------------------+---------------------------+---------------------------+------------------------+-------------+------------+--------+-------+----------+---------------------+
| name | network_acl_id             | vpc_id                    | network_acl_entry_id      | network_acl_entry_name | description | entry_type | policy | port  | protocol | destination_cidr_ip |
+------+----------------------------+---------------------------+---------------------------+------------------------+-------------+------------+--------+-------+----------+---------------------+
| test | nacl-a2dcnd0dgkyzyydwusa15 | vpc-a2db7sc9m92c8gu3a1qw8 | nae-a2dlyltk3rtsjo15du0nf |                        |             |            | accept | -1/-1 | all      | 0.0.0.0/0           |
+------+----------------------------+---------------------------+---------------------------+------------------------+-------------+------------+--------+-------+----------+---------------------+
```

### Count of network ACLs per region

```sql
select
  region,
  count(*) network_acl_count
from
  alicloud_vpc_network_acl
group by
  region;

+------------+-------------------+
| region     | network_acl_count |
+------------+-------------------+
| ap-south-1 | 1                 |
+------------+-------------------+
```

### CODE REVIEW

File: alicloud/plugin.go [hunk 1]
```diff
@@ -31,6 +31,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"alicloud_ram_user":                tableAlicloudRAMUser(ctx),
 			"alicloud_vpc":                     tableAlicloudVpc(ctx),
 			"alicloud_vpc_vswitch":             tableAlicloudVpcVSwitch(ctx),
+			"alicloud_vpc_network_acl":         tableAlicloudVpcNetworkAcl(ctx),
```
Comment (Subhajit97) [inline alicloud/plugin.go:34]: Sort the list

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+)
+
+func tableAlicloudVpcNetworkAcl(ctx context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "alicloud_vpc_network_acl",
+		Description: "VPC network ACL.",
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:19]:
```suggestion
		Description: "Alicloud VPC Network ACL.",
```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+			{
+				Name:        "creation_time",
+				Type:        proto.ColumnType_TIMESTAMP,
+				Description: "The time when the network ACL was created.",
+			},
+			{
+				Name:        "vpc_id",
+				Type:        proto.ColumnType_CIDR,
+				Transform:   transform.FromField("Ipv6CidrBlock"),
+				Description: "The ID of the associated VPC.",
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:62]: Fix `vpc_id` column
```suggestion
				Name:        "vpc_id",
				Type:        proto.ColumnType_STRING,
				Description: "The ID of the associated VPC.",
```

Comment (Subhajit97) [reply alicloud/table_alicloud_vpc_network_acl.go:62]: image

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        proto.ColumnType_JSON,
+				Description: "The associated resources.",
+				Transform:   transform.FromField("Resources.Resource"),
+			},
+			{
+				Name:        "owner_id",
+				Type:        proto.ColumnType_STRING,
+				Description: "The ID of the owner of the VPC.",
+			},
+			// steampipe standard columns
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:92]:
```suggestion
			
			// steampipe standard columns
```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+				Name:        "resources",
+				Type:        proto.ColumnType_JSON,
+				Description: "The associated resources.",
+				Transform:   transform.FromField("Resources.Resource"),
+			},
+			{
+				Name:        "owner_id",
+				Type:        proto.ColumnType_STRING,
+				Description: "The ID of the owner of the VPC.",
+			},
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:91]: Add `tags_src` column

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+				Description: ColumnDescriptionTitle,
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getNetworkAclAka,
+				Transform:   transform.FromValue(),
+			},
+			// alicloud standard columns
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:112]:
```suggestion
			
			// alicloud standard columns
```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+			{
+				Name:        "tags",
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("Tags.Tag"),
+				Description: ColumnDescriptionTags,
+			},
+			{
+				Name:        "title",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("NetworkAclName"),
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:102]: Add `NetworkAclId` as fallback value for `title`, if `NetworkAclName` is null.

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+			{
+				Name:        "name",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("NetworkAclName"),
+				Description: "The name of the network ACL.",
+			},
+			{
+				Name:        "network_acl_id",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("NetworkAclId"),
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:39]:
```suggestion
```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+				Name:        "account_id",
+				Description: "The alicloud Account ID in which the resource is located.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getCommonColumns,
+				Transform:   transform.FromField("AccountID"),
+			},
+		},
+	}
+}
+
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:129]:
```suggestion

//// LIST FUNCTION

```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+	client, err := VpcService(ctx, d, region)
+	if err != nil {
+		plugin.Logger(ctx).Error("alicloud_vpc_network_acl.listNetworkAcl", "connection_error", err)
+		return nil, err
+	}
+	request := vpc.CreateDescribeNetworkAclsRequest()
+	request.Scheme = "https"
+	request.PageSize = requests.NewInteger(50)
+	request.PageNumber = requests.NewInteger(1)
+	count := 0
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:141]:
```suggestion
	
	count := 0
```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+	for {
+		response, err := client.DescribeNetworkAcls(request)
+		if err != nil {
+			plugin.Logger(ctx).Error("alicloud_vpc_network_acl.listNetworkAcl", "query_error", err, "request", request)
+			return nil, err
+		}
+		for _, NetworkAcl := range response.NetworkAcls.NetworkAcl {
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:148]: Use lowerCamelCase
```suggestion
		for _, networkAcl := range response.NetworkAcls.NetworkAcl {
```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+		totalcount, err := strconv.Atoi(response.TotalCount)
+		pageNumber, err := strconv.Atoi(response.PageNumber)
+		if count >= totalcount {
+			break
+		}
+		request.PageNumber = requests.NewInteger(pageNumber + 1)
+	}
+	return nil, nil
+}
+
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:162]:
```suggestion

//// HYDRATE FUNCTIONS

```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+	request := vpc.CreateDescribeNetworkAclAttributesRequest()
+	request.Scheme = "https"
+	request.NetworkAclId = id
+
+	response, err := client.DescribeNetworkAclAttributes(request)
+	if serverErr, ok := err.(*errors.ServerError); ok {
+		if serverErr.ErrorCode() == "InvalidNetworkAcl.NotFound" {
+			plugin.Logger(ctx).Warn("alicloud_vpc_network_acl.getNetworkAclAttribute", "not_found_error", serverErr, "request", request)
+			return nil, nil
+		}
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:187]: Use `ShouldIgnoreError` for `notFound` cases.

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+		plugin.Logger(ctx).Error("alicloud_vpc_network_acl.getNetworkAclAttribute", "query_error", serverErr, "request", request)
+		return nil, serverErr
+	}
+	return response.NetworkAclAttribute, nil
+}
+
+func getNetworkAclAka(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getEcsDiskAka")
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:195]:
```suggestion
	plugin.Logger(ctx).Trace("getNetworkAclAka")
```

File: alicloud/table_alicloud_vpc_network_acl.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+				Description: "The name of the region where the resource resides.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("RegionId"),
+			},
+			{
+				Name:        "account_id",
+				Description: "The alicloud Account ID in which the resource is located.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getCommonColumns,
+				Transform:   transform.FromField("AccountID"),
```
Comment (Subhajit97) [inline alicloud/table_alicloud_vpc_network_acl.go:124]: Do we need `getCommonColumns` for `account_id`?
Or, we can use the value of `owner_id` column?

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
Make suggested changes, and also add docs and tests.

**Comment by cbruno10** [comment]
Closing this PR for now (but keeping branch), as this table is still under development.

**Review by Subhajit97 — DISMISSED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**